### PR TITLE
Menu items

### PR DESCRIPTION
**Pull request by loafofpiecrust**
I have been hacking on keymaps for the past week. I was happy to see the addition of `which-key-add-keymap-based-replacements` for command descriptions, but I was bothered by the fact that the implementation needed to make extra bindings when we can use the built-in menu-item syntax for `define-key`. You support this with `which-key-enable-extended-define-key`, but I don't think advice is necessary, as noted in #261 

This PR adds support for built-in menu item entries in keymaps.
Instead of using `describe-buffer-bindings` for the active bindings, we can grab them using `current-active-maps` and filtering by current prefix. Then, the full binding is given, letting us support the full extent of menu items. That means we no longer need pseudo-bindings because we can use the command replacement embedded in the original keymap. In addition, we can fully support both single command menu items and sub-menus.

These changes also reduce the total code footprint by making use of `which-key--get-keymap-bindings` for all `which-key` views, not just ones specific to one keymap.
To add descriptions, we no longer need `which-key-enable-extended-define-key` so I removed it and the associated advice.

Please let me know what you think. I think the design is sound, and I hope my implementation is too. I have used it for the past day or so and have not run into many issues. It also doesn't seem to incur any more overhead; when I profile, the slowest part of `which-key` init by far is the regex replacements.

Logistical note: I haven't done the FSF copyright assignment yet so I'll get on that if this PR is appealing to merge.

### CODE REVIEW

**Comment by justbur** [comment]
Thank you. I looked quickly. My two main concerns with parsing keymaps directly are

- Do you handle the precedence rules correctly when multiple active maps define the same binding?

- What are the performance implications?

Also I think the motivation is not quite clear to me. Pseudo keys are an alternative method for defining keys for which-key to use. People have been using menu-items as a hack to make dynamic bindings, which is not the same.

**Comment by loafofpiecrust** [comment]
- Precedence: I think that I am handling that right. Keymap precedence is defined in the Emacs docs like so `Usually, the active keymaps are: (i) the keymap specified by the keymap property, (ii) the keymaps of enabled minor modes, (iii) the current buffer’s local keymap, and (iv) the global keymap, in that order. Emacs searches for each input key sequence in all these keymaps. `. Testing `(current-active-maps t)` seems to return in that order as well and we're pulling the first instance of each key binding from there. I will have to do more testing to make sure that it covers multiple definitions of a prefix key and edge cases, but I give a provisional yes.
- Performance: Here's a performance trace for my leader key with regex replacements turned on
```
388  77% - timer-event-handler
381  76%  - apply
326  65%   - which-key--update
326  65%    - which-key--create-buffer-and-show
295  58%     - which-key--get-bindings
280  55%      - which-key--format-and-replace
150  29%       + which-key--maybe-replace
 95  18%       + which-key--propertize-description
 26   5%       + which-key--maybe-add-docstring
  7   1%       + kbd
  1   0%         which-key--extract-key
 10   1%      - which-key--get-current-bindings
  8   1%       - mapcan
  8   1%        - #<compiled 0x19f099e583673454>
  8   1%         - which-key--get-keymap-bindings
  8   1%          - let
  8   1%           + map-keymap
  1   0%       + cl-remove-if-not
  1   0%       + cl-remove-duplicates
  4   0%      + sort
 30   5%     + which-key--create-pages
```

And a performance trace for my leader key with regex replacements turned *off* (since they're the biggest performance hit):
```
717  85% - timer-event-handler
714  85%  - apply
551  66%   - which-key--update
550  65%    - which-key--create-buffer-and-show
388  46%     - which-key--get-bindings
359  43%      + which-key--format-and-replace
 21   2%      - which-key--get-current-bindings
 16   1%       - mapcan
 16   1%        - #<compiled 0x19f099e583673454>
 16   1%         + which-key--get-keymap-bindings
  5   0%       + cl-remove-duplicates
  8   0%      + sort
162  19%     + which-key--create-pages
  1   0%      which-key--this-command-keys
```

I don't notice extra delays and from these CPU traces, it looks like `which-key--get-current-bindings` likely isn't causing slowdowns.

- Motivation: using the built-in menu items mechanism means that the which-key buffer could theoretically/eventually align with the menu-bar for a particular mode. This is appealing to me for UX purposes and because it means that when I look at my keymaps in `help` buffers they aren't littered with `<which-key>` bindings. Based on just reading the [docs for menu items](https://www.gnu.org/software/emacs/manual/html_node/elisp/Simple-Menu-Items.html#Simple-Menu-Items) putting a string name in there is not a hack at all. The [conversation here](https://github.com/oantolin/embark/issues/50) even demonstrates a way in which you can get a super basic binding hint with stock Emacs by giving the keymap a string name and simply doing `(define-key map "f" '("Find file!" . find-file))`. With this PR, `which-key` would be compatible with that methodology, though not yet as snappy!

**Comment by minad** [comment]
@loafofpiecrust @justbur I would very much love to see menu-item support in which-key! See also the feature request https://github.com/justbur/emacs-which-key/issues/177.

**Comment by loafofpiecrust** [comment]
Be aware that I have not implemented support for any menu item properties, like the conditional items in the issue @minad references, just the description field from simple and expanded menu items. I figured that's a good place to start and supporting any other properties can be evaluated later on.

**Comment by minad** [comment]
@loafofpiecrust For the use cases I have in mind, the simple menu items would be totally sufficient!

**Comment by justbur** [comment]
@loafofpiecrust ok, I've come around to the idea, and I'm open to it. If you're willing to support issues that might arise, I'm willing to merge it after going through the code. 

The idea behind the extended key definition stuff was to create a temporary workaround for emacs not having a simple way to expose the `(string . def)` bindings through something akin to `describe-buffer-bindings`. I actually wrote a patch at one point for emacs to do this, and the plan was to remove the advice when that patch became available. The process of getting that patch accepted stalled though, and I got busy and lost track of it. 

I do think having which-key just read the keymaps and know about `(string . def)` as well as menu-items is attractive. However, I was hesitant about the two issues I mentioned to you and didn't have the time to really investigate myself. 

On the performance side, my understanding is that the profiler is not very good at measuring performance (it's also hard to interpret what you're showing, since you're only showing the calls that come out of `timer-event-handler`). If which-key starts to feel sluggish I think that would ruin the experience. That doesn't mean, I know there will be performance issues. It's just that as you're aware this code is run very often when which-key is enabled. 

> @loafofpiecrust For the use cases I have in mind, the simple menu items (string . def) would be totally sufficient!

@minad That's exactly the point of the keymap-based replacement, which is already available (see [here](https://github.com/justbur/emacs-which-key#keymap-based-replacement))

**Comment by minad** [comment]
@justbur

> That's exactly the point of the keymap-based replacement, which is already available (see here)

Yes, but I would just like to manipulate keymaps directly instead of bothering with the which-key pseudo keys. I would very much prefer to use the Emacs standard (string . def) menu items instead!

**Comment by loafofpiecrust** [comment]
@minad Thanks for the further explanation! A patch to `describe-buffer-bindings` might have been the way to go, but at least this method doesn't involve parsing the raw string. If there's a performance or other issue that can't be solved outside of core, then I might be inclined to look at `current-active-maps` for a patch or exposing the logic behind `describe-buffer-bindings` as another function.

I have been testing this branch with `which-key-idle-delay` down to `0.1` and don't notice an added delay, nor do I with secondary maps (pressing `SPC`, waiting a second, pressing `p`). I just pushed a commit optimizing it a bit, but the larger slowdowns that I have noticed in `which-key` are regex replacements (in DOOM there's a big fat list of them) and `(documentation xxx)` with key substitutes which I disabled to get an idea of this branch's performance impact alone.

Lastly, I am definitely willing to support with issues that come up!
The only glaring one that I've noticed so far is `<override-state> -> all` from `evil` shows up. I need to check for that and make it ignore following keymaps if one has `<override-state>` (I think that's how that is supposed to work!).

**Comment by loafofpiecrust** [comment]
I now have the FSF copyright assignment! I will add one or two more commits to address some single keymaps not displaying right, but this PR is generally ready for code review.

**Comment by minad** [comment]
@justbur @loafofpiecrust Is anything missing here? From my perspective this sounds like a great addition. And after looking at the code, it also seems like a solid approach, in particular since it gets rid of the describe-buffer-bindings parsing. Removing the advice on define-key is also a good idea!

File: which-key.el [hunk 1]
```diff
@@ -1811,114 +1770,89 @@ ones. PREFIX is for internal use and should not be used."
...
+         (keys-list (listify-key-sequence prefix))
+         (get-bindings (lambda (map)
+                         (which-key--get-keymap-bindings map nil nil keys-list)))
+         (comparator (lambda (a b)
+                       (string= (car a) (car b))))
+         ;; Get bindings from all active keymaps that match the given prefix.
+         (all-keymaps (mapcar get-bindings (current-active-maps t)))
+         ;; If one of those maps binds <override-state>, then it should shadow others.
+         (binding-is-overriding (lambda (binding)
+                                  (string-suffix-p "<override-state>" (car binding))))
```
Comment (justbur) [inline which-key.el:1838]: I'm not sure about the precedence between evil's override state and the overriding keymaps in emacs. I believe the latter has higher preference.

File: README.org [hunk 2]
```diff
@@ -265,8 +265,24 @@
 **** Keymap-based replacement
      Using this method, which-key can display a custom string for a key
      definition in some keymap. There are two ways to define a keymap-based
-     replacement. The first is to use
-     =which-key-add-keymap-based-replacements=. The statement
+     replacement. The first is to define menu items,
+     [[https://www.gnu.org/software/emacs/manual/html_node/elisp/Simple-Menu-Items.html#Simple-Menu-Items][simple]]
+     or
+     [[https://www.gnu.org/software/emacs/manual/html_node/elisp/Extended-Menu-Items.html#Extended-Menu-Items][extended]],
```
Comment (justbur) [inline README.org:271]: How about some examples of extended menu items for the readme?

**Review by justbur — CHANGES_REQUESTED** [review]
Looks pretty good. I added some comments.

File: which-key.el [hunk 3]
```diff
@@ -525,24 +525,6 @@ it."
   :group 'which-key
   :type 'boolean)
 
-(defcustom which-key-enable-extended-define-key nil
```
[the remark this answers is not in the document]
Comment (loafofpiecrust) [reply which-key.el:528]: Looking at the code again, this variable could still toggle the use of menu item replacements (default to `t`) so that a user could toggle between showing command name vs replacement string on the fly. Thoughts on that? If that doesn't sound good I'll just do `make-obsolete-variable` as you suggest.

File: which-key.el (hunk 1 above)
Comment (loafofpiecrust) [reply which-key.el:1838]: Yeah doing a bit more reading, it looks like evil override-state is intended just to override global evil bindings, not any other keymap. So I'm not exactly sure what logic to use here, but probably not what I implemented so far.

File: which-key.el [hunk 4]
```diff
@@ -933,12 +915,18 @@ both have the same effect for the \"C-x C-w\" key binding, but
...
-        `(which-key ,(cons string command))))
+    (let* ((string (if (stringp replacement)
+                       replacement
+                     (car-safe replacement)))
+           (key-internal (kbd key))
+           (command (lookup-key keymap key-internal))
+           ;; If a command was specified, check that against the existing binding.
+           (command-verified (or (null (cdr-safe replacement))
+                                 (equal command (cdr-safe replacement)))))
+      ;; Only bind to symbols, since a number indicates an error in lookup-key.
```
[the remark this answers is not in the document]
Comment (jhenahan) [reply which-key.el:926]: Data point: I use nil bindings extensively with `general.el` for prefix keys, so I'd hate to lose that unless there's a better way to do that.

[the remark this answers is not in the document]
Comment (loafofpiecrust) [reply which-key.el:943]: Yeah great point, I've fixed this in my most recent commits, so that binding descriptions on nil are allowed.

File: README.org (hunk 2 above)
Comment (loafofpiecrust) [reply README.org:271]: I added a small example of each in a recent commit, do those look sufficient?

Comment (justbur) [reply README.org:271]: yes, thank you

File: which-key.el (hunk 4 above)
[the remark this answers is not in the document]
Comment (justbur) [reply which-key.el:943]: It's not working for me. I should be able to do something like

```elisp
(define-key map "b" '("buffers"))
(define-key map "bb" 'switch-buffer)
```

to name the "b" prefix in `map` for which-key.

[the remark this answers is not in the document]
Comment (justbur) [reply which-key.el:943]: In case it's not clear, this works in the current version.

File: which-key.el (hunk 1 above)
Comment (bdarcus) [reply which-key.el:1856]: Great to see this PR. Is this the last remaining issue?

**Comment by a13** [comment]
any news here?

**Comment by bdarcus** [comment]
> any news here?

@loafofpiecrust - any update? Would be great to get this merged.

**Comment by justbur** [comment]
See #308. I decided to do it myself. If you are interested in this feature. Please test that PR and report there whether you run into bugs.

**Comment by justbur** [comment]
I merged support for menu-items

**Comment by loafofpiecrust** [comment]
Thank you for carrying this change home @justbur, stuff with my job and living situation got really hectic this past month.

**Comment by justbur** [comment]
@loafofpiecrust no problem. Thanks for your work on this
